pkg/ingester/limiting.go

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingester
+
+import (
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/shirou/gopsutil/v3/process"
+	"github.com/weaveworks/common/httpgrpc"
+)
+
+func (i *Ingester) updateResourceUtilization() {
+	if !i.cfg.UtilizationBasedLimitingEnabled {
+		return
+	}
+
+	lastUpdate := i.lastResourceUtilizationUpdate.Load()
+
+	proc, err := process.NewProcess(int32(os.Getpid()))
+	if err != nil {
+		level.Warn(i.logger).Log("msg", "couldn't update CPU/memory utilization - failed to get process info",
+			"err", err.Error())
+		return
+	}
+
+	memInfo, err := proc.MemoryInfo()
+	if err != nil {
+		level.Warn(i.logger).Log("msg", "couldn't update CPU/memory utilization - failed to get memory info",
+			"err", err.Error())
+		return
+	}
+	i.memoryUtilization.Store(memInfo.RSS)
+
+	now := time.Now().UTC()
+
+	cput, err := proc.Times()
+	if err != nil {
+		level.Warn(i.logger).Log("msg", "couldn't update CPU/memory utilization - failed to get process times",
+			"err", err.Error())
+		return
+	}
+
+	lastCPUTime := i.lastCPUTime.Load()
+	// TODO: Should Guest and GuestNice be subtracted, as potentially duplicated in User?
+	// https://github.com/mackerelio/go-osstat/blob/54b9216143f9aa087151258f103dbd5c381f7915/cpu/cpu_linux.go#L70-L74
+	cpuTime := cput.User + cput.System + cput.Idle + cput.Nice + cput.Iowait + cput.Irq +
+		cput.Softirq + cput.Steal + cput.Guest + cput.GuestNice
+	i.lastCPUTime.Store(cpuTime)
+
+	i.lastResourceUtilizationUpdate.Store(now)
+
+	if lastUpdate.IsZero() {
+		return
+	}
+
+	cpuUtil := (cpuTime - lastCPUTime) / now.Sub(lastUpdate).Seconds()
+	i.cpuUtilization.Store(cpuUtil)
+
+	level.Info(i.logger).Log("msg", "resource utilization", "memory_utilization", memInfo.RSS,
+		"cpu_utilization", cpuUtil)
+}
+
+// checkReadOverloaded checks whether the ingester read path is overloaded wrt. CPU and/or memory.
+func (i *Ingester) checkReadOverloaded() error {
+	if !i.cfg.UtilizationBasedLimitingEnabled {
+		return nil
+	}
+
+	memUtil := i.memoryUtilization.Load()
+	cpuUtil := i.cpuUtilization.Load()
+	lastUpdate := i.lastResourceUtilizationUpdate.Load()
+	if lastUpdate.IsZero() {
+		return nil
+	}
+
+	memPercent := 100 * (memUtil / i.readPathMemoryThreshold)
+	cpuPercent := 100 * (cpuUtil / i.readPathCPUThreshold)
+
+	level.Info(i.logger).Log("msg", "resource utilization",
+		"memory_threshold", i.readPathMemoryThreshold, "memory_percentage", memPercent,
+		"cpu_threshold", i.readPathCPUThreshold, "cpu_percentage", cpuPercent)
+
+	if memPercent >= 100 {
+		return httpgrpc.Errorf(http.StatusTooManyRequests, "the ingester is currently too busy to process queries, try again later")
+	}
+	if cpuPercent >= 100 {
```

Comment:
Since CPU usage can be very flappy, where it might go above our threshold for a millisecond then back down to near zero for a long time, we might want to guard against that so that we don't trigger request rejections until we're reliably above the threshold. For that, we might use some kind of average over a time period to threshold off of rather than instantaneous values. For example, we could use sliding window average, or an exponential weighted moving average over the last N seconds.

Reply:
I've implemented EWMA calculation of the CPU usage.